src/solver.rs

```diff
@@ -8,8 +12,99 @@ pub fn query_smt(query: &String) -> bool {
 }
 
 pub fn query_smt_with_solver(query: &String, solver: SolverConfig) -> bool {
-    let query = format!("{}\n{}", query, "(check-sat)");
+    let (output, stderr) = query_smt_internal(&format!("{query}\n(check-sat)\n"), solver);
 
+    match output.as_str() {
+        "sat\n" => true,
+        "unsat\n" => false,
+        _ => panic!(
+            "Invalid output from smt solver.\nQuery: {query}\nStdout: {output}\nStderr: {stderr}",
+        ),
+    }
+}
+
+/// A value in the model, either as a parsed number or an unparsed s-expression.
+#[derive(Debug, PartialEq)]
+pub enum ModelValue {
+    Number(BigUint),
+    String(String),
+}
+
+impl Display for ModelValue {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ModelValue::Number(n) => write!(f, "{n}"),
+            ModelValue::String(s) => write!(f, "{s}"),
+        }
+    }
+}
+
+pub fn query_smt_with_solver_and_eval(
+    query: &String,
+    exprs_to_eval: &Vec<String>,
+    solver: SolverConfig,
+) -> (bool, Vec<ModelValue>) {
+    let mut sat_query = format!("{query}\n(check-sat)\n");
+    if !exprs_to_eval.is_empty() {
+        sat_query = format!(
+            "(set-option :produce-models true)\n{sat_query}(get-value ({}))\n",
+            &exprs_to_eval.join(" ")
+        )
+    }
+    let (output, stderr) = query_smt_internal(&sat_query, solver);
+    let lines = output.split('\n').collect::<Vec<_>>();
```

Comment:
maybe also filter out empty strings after this? Some solvers might have a blank line between the result and the model